File: scope104/src/main/java/com/petro/scope104/data/network/UserDataSourceNetworkImpl.java
package com.petro.scope104.data.network;

import androidx.annotation.NonNull;


import com.petro.scope104.data.UserDataSource;
import com.petro.scope104.data.network.response.UserListResponse;
import com.petro.scope104.domain.entity.WorkerEntity;
import com.petro.scope104.presentation.list.Gender;

import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Observable;

public class UserDataSourceNetworkImpl implements UserDataSource {
    private final GitHubService service;
    private final NetworkConverter networkConverter;
    @Inject
    public UserDataSourceNetworkImpl(GitHubService service, NetworkConverter networkConverter) {
        this.service = service;
        this.networkConverter = networkConverter;
    }

    @Override
    public Observable<List<WorkerEntity>> loadUsers(int pageNumber, int pageSize, @NonNull Gender gender, @NonNull List<String> countries) {
        final Observable<UserListResponse> apiResponseLiveData = service.listRepos(String.join(",", countries), pageNumber, pageSize, gender.serverName);
        return apiResponseLiveData.map(input -> input.results.stream().map(networkConverter::mapNetworkToUi).collect(Collectors.toList()));
    }

    @Override
    public void saveUsers(List<WorkerEntity> users) {
        throw new IllegalStateException("Server not found!");
    }
}

File: scope104/src/main/kotlin/com/petro/scope104/data/network/UserDataSourceNetworkImpl.kt
package com.petro.scope104.data.network

import com.petro.scope104.data.UserDataSource
import com.petro.scope104.data.network.response.UserListResponse
import com.petro.scope104.data.network.response.UserResponse
import com.petro.scope104.domain.entity.WorkerEntity
import com.petro.scope104.presentation.list.Gender
import io.reactivex.rxjava3.core.Observable
import java.util.stream.Collectors
import javax.inject.Inject

class UserDataSourceNetworkImpl @Inject constructor(
    private val service: GitHubService,
    private val networkConverter: NetworkConverter
) : UserDataSource {
    override fun loadUsers(
        pageNumber: Int,
        pageSize: Int,
        gender: Gender,
        countries: List<String>
    ): Observable<List<WorkerEntity>> {
        val apiResponseLiveData: Observable<UserListResponse> = service.listRepos(
            java.lang.String.join(",", countries),
            pageNumber,
            pageSize,
            gender.serverName
        )
        return apiResponseLiveData.map<List<WorkerEntity>> { input: UserListResponse ->
            input.results.stream()
                .map { user: UserResponse -> networkConverter.mapNetworkToUi(user) }
                .collect(Collectors.toList())
        }
    }

    override fun saveUsers(users: List<WorkerEntity>) {
        throw IllegalStateException("Server not found!")
    }
}